### DISCUSSION TITLE
How to resolve [ReferenceError: useNuxtApp is not defined] in component tests?

### DISCUSSION THREAD

**Discussion by Vivi-Ornitier** [nuxt/nuxt#17060] [Questions]
How to resolve [ReferenceError: useNuxtApp is not defined] in component tests?

I'm using custom plugin by `const { customPlugin } = useNuxtApp()`.
When I run tests for that component, I get ReferenceError: useNuxtApp is not defined.
How can I resolve this issue?

**Comment by huang-julien** [comment]
with vitest i usually do 
`````ts
vi.mock('#imports', () => {
  return {
    useState: (str, init) => {
      if (!cache[str]) {
        cache[str] = { value: init() }
      }
      return cache[str]
    },
    onMounted: cb => cb()
  }
})
````````
tests won't auto import auto injected imports.

**Reply by Vivi-Ornitier** [reply]
@huang-julien 

Thank you for reply.
I tried it.

```
vi.mock("#imports", () => {
  return {
    $customPlugin: (value: string) => value,
  };
});
```

but I get the following error:

`TypeError: Cannot destructure property '$customPlugin' of 'useNuxtApp(...)' as it is undefined.`

Can you give me a more detailed solution?

**Comment by huang-julien** [comment]
what i mocked in my comment above was `useState` and `onMounted`. What you want to do is to mock `useNuxtApp()`, not `customPlugin()` :

```ts
const useNuxtAppMock = {
    $customPlugin: (value: string) => value,
}
vi.mock("#imports", () => {
  return {
    useNuxtApp: () =>useNuxtAppMock
  };
});
```

Also be aware that imports order are very important in your test files when you are mocking.
https://vitest.dev/api/#vi-mock
